Fix linking Athena with VecGeom+CUDA

pcanal opened the pull request:
`libvecgeomcuda.[a|so]` is the final library and it includes the result
of nvlink.  It is not appropriate to use as the intermediary link library.

VecGeom also never produce a shared 'middle' library, so we just always
need to use`libvecgeomcuda.[a|so]` is the final library and it includes the result
of nvlink.  It is not appropriate to use as the intermediary link library.

VecGeom also never produce a shared 'middle' library, so we just always
need to use `libvecgeomcuda_static.a`

Other fixes includes:

* removing spurrious dependencies
* adding support for private linking
* fix handling of some internals when seeing static libraries.

pcanal committed:
Fix VecGeom middle library name.

`libvecgeomcuda.[a|so]` is the final library and it includes the result
of nvlink.  It is not appropriate to use as the intermediary link library.

VecGeom also never produce a shared 'middle' library, so we just always
need to use `libvecgeomcuda_static.a`

For the static build this is perfect.
For the shared build (especially in conjunction with user's own
package shared build), this is not working.  So we just keep
using libvecgeomcuda.so as the middle library hoping that the
nvlink result is not pickup at the end.

pcanal committed:
cuda_rdc_target_link_libraries: fix handling of static library dependency.

The previous code was inaccurately looking at the target rather than its dependency when making the choice on how
to place the library

pcanal committed:
CudaRdc: add/fix support for private dependency.

They are record as '$<LINK_ONLY:(.*)>' which was not unbox to replace the cuda libraries
with their 'middle' counterpart.

sethrj reviewed (changes requested):
Currently fails for me on a `+shared~geant4+vecgeom ^vecgeom+shared` build on wildstyle:
```
[906/1015] Linking CXX executable bin/orange-update
FAILED: bin/orange-update
: && /usr/bin/c++ -Wall -Wextra -pedantic -fdiagnostics-color=always -O3 -DNDEBUG  app/CMakeFiles/orange-update.dir/orange-update.cc.o -o bin/orange-update  -Wl,-rpath,/home/s3j/Code/celeritas/build-temp-nog4/lib64:/projects/spack/opt/spack/gcc-8.5.0/vecgeom/ykhz7jg/lib64:  lib64/liborange_final.so  lib64/liborange.so  lib64/libcorecel.so  -Wl,-rpath-link,/home/s3j/Code/celeritas/build-temp-nog4/lib64:/projects/spack/opt/spack/gcc-8.5.0/vecgeom/ykhz7jg/lib64 && :
/projects/spack/opt/spack/gcc-8.5.0/vecgeom/ykhz7jg/lib64/libvecgeom.so: undefined reference to `void vecgeom::cxx::DevicePtr<vecgeom::cuda::SIMDSpecializedVolImplHelper<vecgeom::cuda::EllipticalTubeImplementation, -1, -1> >::Construct<vecgeom::cxx::DevicePtr<vecgeom::cuda::LogicalVolume>, vecgeom::cxx::DevicePtr<vecgeom::cuda::Transformation3D>, unsigned int, int, int>(vecgeom::cxx::DevicePtr<vecgeom::cuda::LogicalVolume>, vecgeom::cxx::DevicePtr<vecgeom::cuda::Transformation3D>, unsigned int, int, int) const'
```

pcanal commented:
The failure is reproduced locally.  The issue is that RdcCuda does not yet support the case of a `final` shared library with a `static` middle/object library (up to now they have always been the same type).  I am exploring 2 options.  Either update VecGeom to produce a 3rd library (the middle/object library) when building shared libraries or updated RdcCuda cmake fragment to handle the case.

pcanal commented:
The only failure left is:
```
/usr/bin/ld: final link failed: No space left on device
```

sethrj commented:
Um how much duplication is in these?

pcanal commented:
> Um how much duplication is in these?

Well ... now we use vecgeom_static must be copied over I guess ... It can be fixed by changing the VecGeom build.

sethrj commented:
So if we have our main libraries (3 of which indirectly link against vecgeom), 4 celeritas tests, and 5 apps, will we have 12 copies of all the vecgeom device symbols? Or 24? Or much fewer?

pcanal commented:
The problem really comes from the executable with 179 of them being more than 30M ... I have to fix this ....

pcanal commented:
Humm .. actually probably can't do anything about it short of having a non-final but shared `libvecgeomcuda` library.

For comparison the size of executable in the static-static build is 100MB !!!

sethrj commented:
@pcanal Here are the sizes of the different shared+shared builds on wildstyle. The build directory is now 2-3× larger...

<img width="452" alt="Screenshot 2024-11-08 at 08 05 38" src="https://github.com/user-attachments/assets/da540cfe-f3d0-450b-bc10-c27a7cbee322">

pcanal committed:
Use new VecGeom 'middle' library if available.

This is available in VecGeom 1.2.10 and later and is noticeable also by the setting
of the CUDA_RDC property

sethrj commented:
@pcanal Are the file sizes smaller with this build? Will it require the pending vecgeom MR to go in?

pcanal commented:
The size should be back to normal.  This PR does not require the VecGeom update but will take advantage of it as soon as it is available.

pcanal commented:
@sethrj Do you know why the docker tests are being skipped?

sethrj commented:
@pcanal To reduce resource use they're skipped in draft mode. I'll take a look in a minute (build on wildstyle) and then the next push should trigger docker as well.

sethrj commented:
<img width="573" alt="percentages" src="https://github.com/user-attachments/assets/d6ca90d8-f361-46eb-bcdc-609db490e3c6">

Hey, with the changes the reldeb sizes are even smaller! Thanks @pcanal

sethrj commented on cmake/FindVecGeom.cmake:
```diff
@@ -26,21 +26,26 @@ if(VecGeom_FOUND AND TARGET VecGeom::vecgeomcuda)
+     set(_vecgeom_middle_library_suffix "_static")
   else()
      set(_vecgeom_cuda_runtime "Shared")
+     set(_vecgeom_middle_library_suffic "")
```
```suggestion
     set(_vecgeom_middle_library_suffix "")
```

sethrj commented on cmake/FindVecGeom.cmake:
```diff
@@ -51,15 +56,17 @@ if(VecGeom_FOUND AND TARGET VecGeom::vecgeomcuda)
-    )
-  endforeach()
+  if(NOT _vecgeom_lib_rdc_final)
+    # Inform celeritas_add_library code
```
Should be `cuda_rdc_add_library` now?

sethrj reviewed (approved):
Great  work!

pcanal committed:
fix typo in FindVecGeom

pcanal committed:
update name of custom add_library in comment

esseivaju reviewed (approved):
Looks good! Tested within Athena and everything's working
